module/zfs/dmu_send.c

```diff
@@ -2844,24 +2931,30 @@ receive_process_record(struct receive_writer_arg *rwa,
 		err = receive_write_embedded(rwa, drrwe, rrd->payload);
 		kmem_free(rrd->payload, rrd->payload_size);
 		rrd->payload = NULL;
-		return (err);
+		break;
 	}
 	case DRR_FREE:
 	{
 		struct drr_free *drrf = &rrd->header.drr_u.drr_free;
-		return (receive_free(rwa, drrf));
+		err = receive_free(rwa, drrf);
+		break;
 	}
 	case DRR_SPILL:
 	{
 		struct drr_spill *drrs = &rrd->header.drr_u.drr_spill;
 		err = receive_spill(rwa, drrs, rrd->payload);
 		kmem_free(rrd->payload, rrd->payload_size);
 		rrd->payload = NULL;
-		return (err);
+		break;
 	}
 	default:
 		return (SET_ERROR(EINVAL));
 	}
+
+	if (err != 0)
+		dprintf_drr(rrd, err);
```

Comment:
No, our convention is to use `SET_ERROR` for hard-coded constants, so the error gets logged at the point where it is actually generated. With ecb2b7dc7f8f7273c215fb30c984bde08e75b852 we should be getting all those cases now.